File: drivers/sdk/host.py
"""QuarkGate driver host — JSON stdin/stdout IPC from Go edge (Python)."""
from __future__ import annotations

import importlib.util
import json
import sys
import traceback
from pathlib import Path
from typing import Any, Dict


def load_driver(provider: str) -> Any:
    driver_path = Path(__file__).resolve().parent.parent / provider / "driver.py"
    if not driver_path.exists():
        raise FileNotFoundError(f"driver not found for {provider}")
    spec = importlib.util.spec_from_file_location(f"qg_driver_{provider}", driver_path)
    if spec is None or spec.loader is None:
        raise ImportError(f"cannot load {driver_path}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def fail(err: Exception, code: str = "DRIVER_ERROR") -> None:
    print(json.dumps({"error": str(err), "code": code}), file=sys.stderr)
    sys.exit(1)


def main() -> None:
    try:
        input_data: Dict[str, Any] = json.load(sys.stdin)
    except Exception as e:
        fail(e)

    provider = input_data.get("provider")
    if not provider:
        fail(ValueError("provider required"))

    driver = load_driver(provider)
    envelope = input_data.get("envelope")
    if isinstance(envelope, str):
        envelope = json.loads(envelope)

    action = input_data.get("action")
    try:
        if action == "estimate":
            micro = driver.estimate_max_cost(envelope or {"payload": {}})
            print(json.dumps({"estimate_micro": micro}))
            return
        if action == "prepare":
            downstream = driver.prepare_request(
                {
                    "envelope": envelope,
                    "credential": input_data.get("credential"),
                    "baseURL": input_data.get("baseURL"),
                }
            )
            print(json.dumps(downstream))
            return
        if action == "normalize":
            raw = input_data.get("raw_usage") or {}
            if hasattr(driver, "normalize_usage"):
                patched = driver.normalize_usage(raw, envelope)
                print(json.dumps({"raw_usage": patched}))
            else:
                print(json.dumps({"raw_usage": raw}))
            return
        if action == "parseResponse":
            if hasattr(driver, "parse_response"):
                raw = driver.parse_response(
                    {
                        "headers": input_data.get("headers") or {},
                        "body": input_data.get("body") or "",
                        "streaming": bool(input_data.get("streaming")),
                        "envelope": envelope,
                    }
                )
                print(json.dumps({"raw_usage": raw or {}}))
            else:
                print(json.dumps({"raw_usage": {}}))
            return
        if action == "healthCheck":
            if hasattr(driver, "health_check"):
                result = driver.health_check(
                    {
                        "baseURL": input_data.get("baseURL"),
                        "credential": input_data.get("credential"),
                    }
                )
                print(json.dumps(result))
            else:
                print(json.dumps({"ok": True, "latency_ms": 0, "message": "no health_check implemented"}))
            return
        if action == "poll":
            if not hasattr(driver, "poll_run"):
                fail(ValueError("poll not supported"))
            result = driver.poll_run(
                {
                    "baseURL": input_data.get("baseURL"),
                    "credential": input_data.get("credential"),
                    "poll_context": input_data.get("poll_context") or {},
                }
            )
            print(json.dumps(result))
            return
        fail(ValueError(f"unknown action: {action}"))
    except Exception as e:
        traceback.print_exc()
        fail(e)


if __name__ == "__main__":
    main()

File: drivers/sdk/host.js
/**
 * QuarkGate driver host — JSON stdin/stdout IPC from Go edge.
 */
const fs = require('fs');
const path = require('path');

function loadDriver(provider) {
  const jsPath = path.join(__dirname, '..', provider, 'driver.js');
  if (fs.existsSync(jsPath)) {
    return require(jsPath);
  }
  const pyPath = path.join(__dirname, '..', provider, 'driver.py');
  if (fs.existsSync(pyPath)) {
    throw new Error('use host.py for Python drivers');
  }
  throw new Error(`driver not found for ${provider}`);
}

function fail(err, code = 'DRIVER_ERROR') {
  console.error(JSON.stringify({ error: err.message || String(err), code }));
  process.exit(1);
}

async function main() {
  let input;
  try {
    input = JSON.parse(fs.readFileSync(0, 'utf8'));
  } catch (e) {
    fail(e);
  }

  const provider = input.provider;
  if (!provider) {
    fail(new Error('provider required'));
  }

  const driver = loadDriver(provider);
  const envelope = typeof input.envelope === 'string'
    ? JSON.parse(input.envelope)
    : input.envelope;

  switch (input.action) {
    case 'estimate': {
      const micro = driver.estimateMaxCost(envelope || { payload: {} });
      console.log(JSON.stringify({ estimate_micro: micro }));
      return;
    }
    case 'prepare': {
      const downstream = driver.prepareRequest({
        envelope,
        credential: input.credential,
        baseURL: input.baseURL,
      });
      console.log(JSON.stringify(downstream));
      return;
    }
    case 'normalize': {
      const raw = input.raw_usage || {};
      if (typeof driver.normalizeUsage === 'function') {
        const patched = driver.normalizeUsage(raw, envelope);
        console.log(JSON.stringify({ raw_usage: patched }));
      } else {
        console.log(JSON.stringify({ raw_usage: raw }));
      }
      return;
    }
    case 'parseResponse': {
      const headers = input.headers || {};
      const body = input.body || '';
      if (typeof driver.parseResponse === 'function') {
        const raw = driver.parseResponse({
          headers,
          body,
          streaming: Boolean(input.streaming),
          envelope,
        });
        console.log(JSON.stringify({ raw_usage: raw || {} }));
      } else {
        console.log(JSON.stringify({ raw_usage: {} }));
      }
      return;
    }
    case 'healthCheck': {
      if (typeof driver.healthCheck === 'function') {
        const result = await driver.healthCheck({
          baseURL: input.baseURL,
          credential: input.credential,
        });
        console.log(JSON.stringify(result));
      } else {
        console.log(JSON.stringify({ ok: true, latency_ms: 0, message: 'no healthCheck implemented' }));
      }
      return;
    }
    case 'poll': {
      if (typeof driver.pollRun !== 'function') {
        fail(new Error('poll not supported'));
      }
      const result = await driver.pollRun({
        baseURL: input.baseURL,
        credential: input.credential,
        poll_context: input.poll_context || {},
      });
      console.log(JSON.stringify(result));
      return;
    }
    default:
      fail(new Error(`unknown action: ${input.action}`));
  }
}

main().catch((err) => fail(err));
